src/RecursiveIteration.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace Dhii\Iterator;
+
+use Dhii\Util\String\StringableInterface as Stringable;
+use Traversable;
+
+/**
+ * Represents an iteration in a recursive iterator.
+ *
+ * @since [*next-version*]
+ */
+class RecursiveIteration extends AbstractBaseIteration implements RecursiveIterationInterface
+{
+    /**
+     * An array of path segments.
+     *
+     * @since [*next-version*]
+     *
+     * @var array
+     */
+    protected $path;
+
+    /**
+     * Constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|int|null                   $key   The iteration key.
+     * @param mixed                             $value The iteration value.
+     * @param string[]|Stringable[]|Traversable $path  A list of path segments.
+     */
+    public function __construct($key, $value, $path = [])
+    {
+        $this->_setKey($key)
+             ->_setValue($value)
+             ->_setPathSegments($path);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function getPathSegments()
+    {
+        return $this->path;
+    }
+
+    /**
+     * Sets the path segments.
+     *
+     * @since [*next-version*]
+     *
+     * @param string[]|Stringable[]|Traversable $path A list of path segments.
+     *
+     * @return $this
+     */
+    protected function _setPathSegments($path)
+    {
+        $this->path = $path;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function getDepth()
+    {
+        return count($this->getPathSegments()) - 1;
```

Comment:
The root, however, is not something you enter, just like any other node in which you currently are. The path describes a series of nodes *from your current node*. Otherwise, you are describing something like a segment, which is arbitrary. But in your case, it is not arbitrary; it is from *the current node*.

Reply:
I think you may be right, if you only look at how we achieve iteration and what the path segments signify in that regard. Conceptually, I think it is a problem that we cannot solve **in the scope of iteration**. I'll explain:

In the scope of the interface for which we are having this discussion, `RecursiveIterationInterface`, the path segments are a property of the iteration, which in turn describe the path to the element that was iterated over (in other words, will be yielded at some point by `current()`).

---

Iteration will always be initiated on something, be it the first node in a hierarchy of objects or the outside-most array, to give an example.

```
[
    "a": ["b", "c", "d"],
    "e": ["f"]
]
```

In a case such as the above, the "root" can be considered to be the array construct, which has two immediate children. It's key is not known to us. In a recursive algorithm, the value of `a` would eventually be considered a root itself with three elements.

From this point of view, element `b` has a path of `[b]`, and since it is not the root, it has depth of one (equal to path segments). Once you start backtracking, the key of the parent becomes known (`a`) and the path of its children elements must now be relative to the new root (becoming `[a, b]` with a dept of 1). This is what we're doing in `AbstractRecursiveIterator`.

Given that the topmost root key may never be known, and may even not exist, the root cannot be considered a node in this regard. So you're right that when iterating, the root is not something that you enter into. 🆗 

----

Similarly, in an hierarchy made up of objects that hold some value:

```
Node [R] : 5
    Node [A] : 12
        Node [B] : 20
        Node [C] : 28
        Node [D] : 24
    Node [E] : 15
        Node [F] : 29
```

In **traversal**, Node `R` will be visited first. This is case when the key of the root can be known. When **iterating** over `R` however, you lose this information because the iterator will only acknowledge the children.

Once again, consider if the above object hierarchy is but a sub-hierarchy of a larger one. Discarding the root from the path of an element causes you to lose information. If the path of `C` is just `[A, C]`, then when you take the whole super hierarchy into consideration, the actual path is broken.

However note, that this only happens when iterating, not traversing, and can also be avoided. In fact, our abstract iterator implementation does this. When iterating recursively, the path will continue to be "compiled" during backtracking until the only missing path segment is the top-most root. 🆗 

---

These demonstrate the point that you are trying to make. Iterating over a node means iterating over its children, which results in the node being iterated over not being visited. This leads to paths missing the root node, which would make depth equal to the number of path segments, as you said. 🆗 

What I mean is that doing `foreach ($node as $child)` does not yield `$node` as a `$child` at any point in the iteration.

I'm not entirely sure if this is a design flaw, or merely a result of doing iteration in the way we're doing it. Root nodes are treated in the same way are regular nodes in computer science, for consistency.

Perhaps it is wrong to try and apply **traversal** concepts to **iteration**.

In any case, I will change the code such that `depth = number of path segments`. 🎉

Reply:
Yep. For me it's as simple as thinking of a path `A/C` as: "go to A", "go to C". Every "go to" is a segment. 2 segments = depth of 2. Height of 3 ;)

Please proceed.